fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * LuceneSerializer.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.directory;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import org.apache.lucene.codecs.compressing.CompressionMode;
+import org.apache.lucene.codecs.compressing.Compressor;
+import org.apache.lucene.codecs.compressing.Decompressor;
+import org.apache.lucene.store.ByteArrayDataInput;
+import org.apache.lucene.store.ByteBuffersDataOutput;
+import org.apache.lucene.util.BytesRef;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+public class LuceneSerializer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSerializer.class);
+    private static final int ENCODING_ENCRYPTED = 1;
+    private static final int ENCODING_COMPRESSED = 2;
+    private static final byte COMPRESSION_VERSION_FOR_HIGH_COMPRESSION = 0;
+
+    private static class EncodingState {
+        private boolean compressed;
+        private boolean encrypted;
+
+        private EncodingState() {
+            this.compressed = false;
+            this.encrypted = false;
+        }
+
+        private void setCompressed(boolean compressed) {
+            this.compressed = compressed;
+        }
+
+        private void setEncrypted(boolean encrypted) {
+            this.encrypted = encrypted;
+        }
+
+        private boolean isCompressed() {
+            return compressed;
+        }
+
+        private boolean isEncrypted() {
+            return encrypted;
+        }
+    }
+
+    @Nullable
+    public static byte[] encode(@Nullable byte[] data, boolean compress, boolean encrypt) {
+        if (data == null) {
+            return null;
+        }
+        final EncodingState state = new EncodingState();
+        byte[] encoded = new byte[data.length + 1];
+        System.arraycopy(data, 0, encoded, 1, data.length);
```

Comment:
This one can be avoided by making the `ByteBuffersDataOutput` earlier and using it to (optimistically) output the encoding byte and then compress. If the compressed `size` is too large, then it's a little bit more complicated to remake an uncompressed, since there is no rewind support. But the happy path will be better.

Reply:
Fixed.